Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -290,6 +290,72 @@ For white-box source code testing, you'll need a setup similar to the developer'
+
+<img src="Images/Chapters/0x05c/jadx_java_xrefs.png" width="500px"/>
+
+###### Native Code
+
+Similarly to Java analysis, you can also use Ghidra to analyze native libraries and obtain cross references by right clicking the desired function and selecting **Show References to**.
+
+<img src="Images/Chapters/0x05c/ghidra_native_xrefs.png" width="500px"/>
+
+##### API Usage
+
+The Android platform provides many in-built libraries for frequently used functionalities in applications, for example cryptography, Bluetooth, NFC, network or location libraries. Determining the presence of these libraries in an application can give us valuable information about its nature.
+
+For instance, if an application is importing `javax.crypto.Cipher`, it indicates that the application will be performing some kind of cryptographic operation. Fortunately, cryptographic calls are very standard in nature, i.e, they need to be called in a particular order to work correctly, this knowledge can be helpful when analyzing cryptography APIs. For example, by looking for the `Cipher.getInstance` function, we can determine the cryptographic algorithm being used. With such an approach we can directly move to analyzing cryptographic assets, which often are very critical in an application. Further information on how to analyse Android's cryptographic APIs is discussed in the section  "[Android Cryptographic APIs](0x05e-testing-cryptography "Android Cryptographic APIs")".
+
+Similarly, the above approach can be used to determine where and how an application is using NFC. For instance, an application using Host-based Card Emulation for performing digital payments must use the `android.nfc` package. Therefore, a good stating point for NFC API analysis would be to consult the [Android Developer Documentation](https://developer.android.com/guide/topics/connectivity/nfc/hce "Host-based card emulation overview") to get some ideas and start searching for critical functions such as `processCommandApdu` from the `android.nfc.cardemulation.HostApduService` class. 
+
+
+##### Checking Secure Connections
+
+Android requires applications to use secure network connection, like HTTPS, while communicating with the backend server. When evaluating an application it is important to check the network configuration, as often debug environment (less secure) configurations are pushed into final release build by mistake.
+
+Also, many applications to further strengthen the network security, use Certificate Pinning. Certificate pinning not only deters MITM attacks, but also slows down reverse engineering of network traffic. But there are many pitfalls while implementing it correctly, one such common mistake being - not verifying a self-signed certificate used in certificate pinning.
+
+Implementation of secure connections and verifying them can be an intricate process, and it is covered in more detail in "[Android Network APIs](0x05g-testing-network-communication "Android Network APIs")" section.
```

Comment:
- Fixed close repetition of "cert. pinning"
- Not sure about this affirmation ", but also slows down reverse engineering of network traffic". How does this happen?


```suggestion
The implementation and verification of secure connections can be an intricate process and there are numerous aspects to consider. For instance, many applications use other protocols apart from HTTP such as XMPP or plain TCP packets, or perform certificate pinning in an attempt to deter MITM attacks but unfortunately having severe logical bugs in its implementation or an inherently wrong security network configuration.

Remember that in most of the cases, just using static analysis will not be enough and might even turn to be extremely inefficient when compared to the dynamic alternatives which will get much more reliable results (e.g. using an interceptor proxy). In this section we've just slightly touched the surface, please refer to the section "[Basic Network Monitoring/Sniffing](0x05b-Basic-Security_Testing.md#basic-network-monitoringsniffing "Basic Network Monitoring/Sniffing")" in the "Android Basic Security Testing" chapter and also check the test cases in the chapter "[Android Network APIs](0x05g-testing-network-communication "Android Network APIs")".
```

Reply:
Slowing down as one need to bypass/disable cert pinning and then only the traffic can be analyzed using burp suite or other proxy tools. Does that make sense?